### PR TITLE
Add karpenter

### PR DESCRIPTION
**Pull request by janajs**
https://karpenter.sh/docs/getting-started/

### CODE REVIEW

File: eks/karpenter-scaler.tf [hunk 1]
```diff
@@ -0,0 +1,52 @@
+resource "aws_cloudformation_stack" "karpenter" {
+  count      = var.karpenter_enabled ? 1 : 0
+  name = "karpenter-stack"
```
Comment (abhiyerra) [inline eks/karpenter-scaler.tf:3]:
```
name = "${var.environment_name}-karpenter"
```

File: eks/karpenter-scaler.tf [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  parameters = {
+    ClusterName = aws_eks_cluster.cluster.name
+  }
+
+  template_body = file("${path.module}/karpenter.yml")
```
Comment (abhiyerra) [inline eks/karpenter-scaler.tf:9]: Use https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http to get the data
```
data "http" "karpenter_cloudformation" {
  url = "https://karpenter.sh/docs/getting-started/cloudformation.yaml"
}

template_body = http.karpenter_cloudformation.body
```

File: eks/karpenter.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Comment (abhiyerra) [inline eks/karpenter.yaml:1]: Remove this file.

File: eks/karpenter-scaler.tf (hunk 1 above)
Comment (abhiyerra) [reply eks/karpenter-scaler.tf:3]: Run:

```
terraform fmt
```

**Review by abhiyerra — CHANGES_REQUESTED** [review]

File: eks/karpenter-scaler.tf [hunk 4]
```diff
@@ -0,0 +1,52 @@
+resource "aws_cloudformation_stack" "karpenter" {
```
Comment (abhiyerra) [inline eks/karpenter-scaler.tf:1]: Rename the file `karpenter.tf`

File: eks/karpenter-scaler.tf (hunk 2 above)
Comment (janajs) [reply eks/karpenter-scaler.tf:9]: I thought to make static, because it may change for each release

**Review by abhiyerra — CHANGES_REQUESTED** [review]

File: eks/karpenter.tf [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  source                        = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+  version                       = "4.7.0"
+  create_role                   = true
+  role_name                     = "karpenter-controller-${aws_eks_cluster.cluster.name}"
+  provider_url                  = replace(aws_iam_openid_connect_provider.cluster.url, "https://", "") 
+  oidc_fully_qualified_subjects = ["system:serviceaccount:karpenter:karpenter"]
+}
+
+resource "aws_iam_role_policy" "karpenter_contoller" {
+  name = "karpenter-policy-${aws_eks_cluster.cluster.name}"
```
Comment (abhiyerra) [inline eks/karpenter.tf:27]:
```suggestion
  name = "karpenter-policy-${var.environment_name}"
```

File: eks/karpenter.tf [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  count = var.karpenter_enabled ? 1 : 0  
+  name = "KarpenterNodeInstanceProfile-${aws_eks_cluster.cluster.name}"
+  role = aws_iam_role.node.name
+}
+
+module "iam_assumable_role_karpenter" {  
+  source                        = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+  version                       = "4.7.0"
+  create_role                   = true
+  role_name                     = "karpenter-controller-${aws_eks_cluster.cluster.name}"
```
Comment (abhiyerra) [inline eks/karpenter.tf:21]:
```suggestion
  role_name                     = "karpenter-controller-${var.environment_name}"
```

**Review by abhiyerra — CHANGES_REQUESTED** [review]

**Pull request merged**